src/app/recipes/CreateRecipeForm.tsx

```diff
@@ -0,0 +1,53 @@
+"use client";
+
+import { ChangeEvent, useActionState, useState } from "react";
+import { createRecipeAction } from "@/lib/actions/recipes";
+import { Input } from "@/components/ui/input";
+import { Textarea } from "@/components/ui/textarea";
+import { Button } from "@/components/ui/button";
+
+const createDefaultRecipeName = (content: string | null) => {
+  if (!content) return "";
+
+  return content.substring(2, content.indexOf("\n"));
+};
+
+export default function CreateRecipeForm({ source }: { source: Object }) {
```

Comment:
I was passing whole objects (recipes and sources) in as params for these components. I was wondering if we could use the prisma schema to set the type here but then I looked and saw that you only pass in IDs and query for the rest of the data in all yours. Should I do that here? Is the idea to make these components more self contained?